src/workers/tx_sender.rs

```diff
@@ -0,0 +1,136 @@
+use std::time::Duration;
+use std::{collections::HashMap, sync::Arc};
+
+use log::{info, warn};
+
+use solana_client::nonblocking::tpu_client::TpuClient;
+
+use solana_sdk::signature::Signature;
+use tokio::sync::RwLock;
+use tokio::task::JoinHandle;
+
+use crate::{WireTransaction, DEFAULT_TX_RETRY_BATCH_SIZE};
+
+use super::block_listenser::BlockListener;
+
+/// Retry transactions to a maximum of `u16` times, keep a track of confirmed transactions
+#[derive(Clone)]
+pub struct TxSender {
+    /// Transactions queue for retrying
+    enqueued_txs: Arc<RwLock<HashMap<Signature, (WireTransaction, u16)>>>,
+    /// block_listner
+    block_listner: BlockListener,
+    /// TpuClient to call the tpu port
+    tpu_client: Arc<TpuClient>,
+}
+
+impl TxSender {
+    pub fn new(tpu_client: Arc<TpuClient>, block_listner: BlockListener) -> Self {
+        Self {
+            enqueued_txs: Default::default(),
+            block_listner,
+            tpu_client,
+        }
+    }
+    /// en-queue transaction if it doesn't already exist
+    pub async fn enqnueue_tx(&self, sig: Signature, raw_tx: WireTransaction, max_retries: u16) {
+        if max_retries == 0 {
+            return;
+        }
+
+        if !self
+            .block_listner
+            .confirmed_txs
+            .read()
+            .await
+            .contains(&sig.to_string())
+        {
+            info!("en-queuing {sig} with max retries {max_retries}");
+            self.enqueued_txs
+                .write()
+                .await
+                .insert(sig, (raw_tx, max_retries));
+
+            println!("{:?}", self.enqueued_txs.read().await.len());
+        }
+    }
+
+    /// retry enqued_tx(s)
+    pub async fn retry_txs(&self) {
+        let len = self.enqueued_txs.read().await.len();
```

Comment:
I am not sure if we want to retry these transactions at all. I am thinking TPU client will forward the transactions to FANOUT (12) leaders. If it really fails we can just let the client retry these transactions. Actually retry logic could be way to attack lite-rpc. We area actually using RAM to store all the transactions which could be exploited. @mschneider Let me know if you have different opinion.

Reply:
This is still there but I have set an upper limit. Should do. Maybe we can turn this off by default. Can be turned on using command line arguments.